numba/cuda/cudaimpl.py

```diff
@@ -151,36 +133,42 @@ def _get_unique_smem_id(name):
     return "{0}_{1}".format(name, _unique_smem_id)
 
 
-@lower('ptx.smem.alloc', types.intp, types.Any)
-def ptx_smem_alloc_intp(context, builder, sig, args):
-    length, dtype = args
+@lower(cuda.shared.array, types.Integer, types.Any)
```

Comment:
Does this not require a `LiteralInteger`? (same question throughout where `.literal_value` attr in use).

Reply:
I've changed it to use `ImtegerLiteral`. For the variants that expect a `UniTuple` I couldn't see a way to enforce that it is a tuple of literals - is there a way to do that?

Reply:
> I've changed it to use `ImtegerLiteral`.

Thanks.

> For the variants that expect a `UniTuple` I couldn't see a way to enforce that it is a tuple of literals - is there a way to do that?

I don't think there's a way to "spell" that for `@lower` at present, I think it'd have to accept a `UniTuple` and then scan the contents in the function body via the signature. This should really be picked up at type inference time ahead of here, i.e. `numba.cuda.cudadecl.Cuda_shared_array` ought to do a strong validation first.

Reply:
I've added some stronger validation to `cudadecl` and a test that using non-literals results in a `TypeError` now.